CRM/Contribute/BAO/Contribution.php

```diff
@@ -937,6 +939,14 @@ protected static function getRelatedMemberships($contributionID) {
     foreach ($membershipPayments as $membershipPayment) {
       $membershipIDs[] = $membershipPayment['membership_id'];
     }
+    if (empty($membershipIDs) && !empty($contributionRecurID)) {
+      $memberships = civicrm_api3('Membership', 'get', [
+        'contribution_recur_id' => $contributionRecurID,
+      ])['values'];
+      foreach ($memberships as $membership) {
+        $membershipIDs[] = $membership['id'];
+      }
+    }
```

Comment:
In which case Membership is not linked with contribution via membership payment table?

Reply:
Yeah - why would that happen? Is this a new feature you want core to support?

Reply:
I can only think off for some reason via payment processor extension or custom coding it didn't manage to create link via membership payment and you trying to handle those scenarios in core. If yes than shouldn't it be included in upgrade script of extension?(just asking as I am not against to have this handled in core as well).

Also looking at the code it looks like all membership is retrieved using 'contribution_recur_id', what will happen if one of the Membership is 'Cancelled' or 'Deceased', will the later code won't renew those membership? If it will renew than shouldn't we need to filter membership retrievals?